crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -0,0 +1,380 @@
+            let target_id = match func.dfg[*target_value] {
+                Value::Function(id) => id,
+                _ => continue,
+            };
+            literal_functions.remove(&target_id);
+        }
+    }
+    literal_functions
+}
+
+fn create_apply_functions(
+    ssa: &mut Ssa,
+    variants_map: &HashMap<FunctionSignature, Vec<FunctionId>>,
+) -> HashMap<FunctionSignature, ApplyFunction> {
+    let mut apply_functions = HashMap::new();
+    for (signature, variants) in variants_map.iter() {
+        if variants.len() > 1 {
+            let apply_function = create_apply_function(ssa, signature, variants);
+            apply_functions.insert(
+                signature.clone(),
+                ApplyFunction { id: apply_function, dispatches_to_multiple_functions: true },
+            );
+        } else {
+            apply_functions.insert(
+                signature.clone(),
+                ApplyFunction { id: variants[0], dispatches_to_multiple_functions: false },
+            );
+        }
+    }
+    apply_functions
+}
+
+fn function_id_to_field(function_id: FunctionId) -> FieldElement {
+    (function_id.to_usize() as u128).into()
+}
+
+/// Creates an apply function for the given signature and variants
+fn create_apply_function(
+    ssa: &mut Ssa,
+    signature: &FunctionSignature,
+    function_ids: &[FunctionId],
+) -> FunctionId {
+    assert!(!function_ids.is_empty());
+    ssa.add_fn(|id| {
+        let mut function_builder = FunctionBuilder::new("apply".to_string(), id, signature.runtime);
+        let target_id = function_builder.add_parameter(Type::Numeric(NumericType::NativeField));
+        let params_ids =
+            vecmap(signature.parameters.clone(), |typ| function_builder.add_parameter(typ));
+
+        let mut previous_target_block = None;
+        for (index, function_id) in function_ids.iter().enumerate() {
+            let is_last = index == function_ids.len() - 1;
+            let mut next_function_block = None;
+
+            let function_id_constant = function_builder.numeric_constant(
+                function_id_to_field(*function_id),
+                Type::Numeric(NumericType::NativeField),
+            );
+            let condition =
+                function_builder.insert_binary(target_id, BinaryOp::Eq, function_id_constant);
+
+            // If it's not the last function to dispatch, crate an if statement
+            if !is_last {
+                next_function_block = Some(function_builder.insert_block());
+                let executor_block = function_builder.insert_block();
+
+                function_builder.terminate_with_jmpif(
+                    condition,
+                    executor_block,
+                    next_function_block.unwrap(),
+                );
+                function_builder.switch_to_block(executor_block);
+            } else {
+                // Else just constrain the condition
+                function_builder.insert_constrain(condition);
```

Comment:
Is this constraint ever expected to be necessary?

Reply:
The idea is that in case the compiler fails, it's better to have a false constraint than a function executed that wasn't the one you intended to